src/nvim/buffer.c

```diff
@@ -2855,98 +2849,152 @@ build_stl_str_hl (
+  bool prevchar_isflag = true;
+  bool prevchar_isitem = false;
+
+  // out_p is the current position in the output buffer
+  char_u *out_p = out;
+
+  // out_end_p is the last valid character in the output buffer
+  // Note: The null termination character must occur here or earlier,
+  //       so any user-visible characters must occur before here.
+  char_u *out_end_p = (out + outlen) - 1;
+
+
+  // Proceed character by character through the statusline format string
+  // fmt_p is the current positon in the input buffer
+  for (char_u *fmt_p = usefmt; *fmt_p; ) {
     if (curitem == STL_MAX_ITEM) {
-      /* There are too many items.  Add the error code to the statusline
-       * to give the user a hint about what went wrong. */
-      if (p + 6 < out + outlen) {
-        memmove(p, " E541", (size_t)5);
-        p += 5;
+      // There are too many items.  Add the error code to the statusline
+      // to give the user a hint about what went wrong.
+      if (out_p + 5 < out_end_p) {
+        memmove(out_p, " E541", (size_t)5);
+        out_p += 5;
       }
       break;
     }
 
-    if (*s != NUL && *s != '%')
-      prevchar_isflag = prevchar_isitem = FALSE;
+    if (*fmt_p != NUL && *fmt_p != '%') {
+      prevchar_isflag = prevchar_isitem = false;
+    }
 
-    /*
-     * Handle up to the next '%' or the end.
-     */
-    while (*s != NUL && *s != '%' && p + 1 < out + outlen)
-      *p++ = *s++;
-    if (*s == NUL || p + 1 >= out + outlen)
+    // Copy the formatting verbatim until we reach the end of the string
+    // or find a formatting item (denoted by `%`)
+    // or run out of room in our output buffer.
+    while (*fmt_p != NUL && *fmt_p != '%' && out_p < out_end_p)
+      *out_p++ = *fmt_p++;
+
+    // If we have processed the entire format string or run out of
+    // room in our output buffer, exit the loop.
+    if (*fmt_p == NUL || out_p >= out_end_p)
       break;
 
-    /*
-     * Handle one '%' item.
-     */
-    s++;
-    if (*s == NUL)      /* ignore trailing % */
+    // The rest of this loop wil handle a single `%` item.
+    // Note: We increment here to skip over the `%` character we are currently
+    //       on so we can process the item's contents.
+    fmt_p++;
+
+    // Ignore `%` at the end of the format string
+    if (*fmt_p == NUL) {
       break;
-    if (*s == '%') {
-      if (p + 1 >= out + outlen)
+    }
+
+    // Two `%` in a row is the escape sequence to print a
+    // single `%` in the output buffer.
+    if (*fmt_p == '%') {
+      // Ignore the character if we're out of room in the output buffer.
+      if (out_p >= out_end_p)
         break;
-      *p++ = *s++;
-      prevchar_isflag = prevchar_isitem = FALSE;
+      *out_p++ = *fmt_p++;
+      prevchar_isflag = prevchar_isitem = false;
       continue;
     }
-    if (*s == STL_MIDDLEMARK) {
-      s++;
-      if (groupdepth > 0)
+
+    // STL_MIDDLEMARK denotes the separation place between
+    // left and right aligned items.
```

Comment:
"denotes the separation place" => "separates". Then this comment fits on 1 line.